### PR TITLE
Change getMaterialProperty semantics to match getFunction/getPostprocessorValue

### LINKED ISSUE #5218: getMaterialProperty should behave like getFunction..
...and ```getPostprocessorValue```.
All these functions accept a _parameter_ name and the the object which the value of that parameter names. This will allow us to specify default values for material properties

- [x] Have ```getMaterialProperty``` return a **const** reference!
- [x] Add a derived string type ```MaterialPropertyName```. This will allow us to cleanly check if the argument to ```getMaterialProperty``` is an input parameter (of type MaterialPropertyName)
- [ ] Change ```getMaterialProperty``` and add logic to check if a valid parameter exists, if so get its value.
- [ ] Specialize getMaterialProperty<Real> to check if the value evaluates to a Real number. If so: build a ```MaterialProperty``` object initialized to the real value. Store that constant object in ```MaterialPropertyInterface```.
- [ ] Update ```DerivativeMaterialInterface``` to work with the new system.

Assign to me.

### PR DESCRIPTION
**Pull request by dschwen**
This is a fairly large commit that closes #5218. It augments `getMaterialProperty` to allow a user to pass in an input parameter key for an input parameter of type `MaterialPropertyName`. This avoids the common `getMaterialProperty<>(getParam<std::string>("key"))` pattern and replaces it with `getMaterialProperty<>("key")`. The big benefit is that the input parameters of type `MaterialPropertyName` now accept Real numbers for Real material properties. This allows developers to set numerical default values for material properties which are taken if no material property name is provided. It also allows users to supply constant numbers in input files directly where material properties of type real are needed.

There might be some additional application patches necessary (a one line patch is already in Marmot, Grizzly didn't need anything, I'm working on hyrax). I'll work with @snschune tomorrow to get yak and rattlesnake in shape.

I have a few commits in place to use the new semantics all over framework, modules and tests (btw. the default property stuff is covered in the tests now). I have also edited some input files to remove a few GenericConstantMaterial properties and use default material properties instead. I'm sure more can be done (especially on the applications side).

I will add follow on PRs (for new issues) to update the derivative material property system accordingly, but this PR is big enough as it is.

### CODE REVIEW

**Comment by dschwen** [comment]
@amjokisaari: A patch for hyrax is ready, but it is going to be an incompatible change (due to the redefinition of an input parameter with a new default in a class inherited form phase_field).

File: framework/include/utils/InputParameters.h [hunk 1]
```diff
@@ -996,6 +996,28 @@ InputParameters::setParamHelper<FunctionName, int>(const std::string & /*name*/,
   l_value = oss.str();
 }
 
+template<>
+inline
```
Comment (permcody) [inline framework/include/utils/InputParameters.h]: In general I think that the framework has reached the size where we need to start removing inline code. If the compiler chooses to inline, that's fine but these methods are only called a few times during the setup so there is little reason to continue bloating the final executable size.

File: modules/solid_mechanics/src/actions/DomainIntegralAction.C [hunk 2]
```diff
@@ -175,12 +175,12 @@ DomainIntegralAction::~DomainIntegralAction()
 void
 DomainIntegralAction::act()
 {
-  const std::string uo_name("crackFrontDefinition");
-  const std::string ak_base_name("q");
-  const std::string av_base_name("q");
+  const std::string uo_name = "crackFrontDefinition";
```
Comment (permcody) [inline modules/solid_mechanics/src/actions/DomainIntegralAction.C]: I don't think we mind having constructor syntax for non-integral types. The old way is fine here.

**Comment by permcody** [comment]
@dschwen - We missed several applications with the const property:
examples, tutorial, badger, falcon, bison-casl, etc. I know you may not have access to all of these but we'll need to make sure that we have patches for them so it might be a little bit of time since we are all tied up at the conference. Perhaps we can focus on getting Mike's PR in today and then see if we can get this one in later, maybe tomorrow or Friday?

File: framework/include/utils/InputParameters.h (hunk 1 above)
Comment (dschwen) [inline framework/include/utils/InputParameters.h]: Uhm, @permcody, ```inline``` in C++ is not the same thing as ```inline``` in C. This statement has _no_ impact on the executable size as far as I know. It is necessary to keep the full template specialization in the header file. Omit it and you get duplicate symbol errors. I just went with how it was done in that file. In ```MaterialPropertyInterface``` I chose to only forward declare the full specializations in the header (necessary so the compiler knows about them, damn that bug cost me a good half hour) and put the implementations in the source file. Full specializations are not templates anymore. It should improve compile times if we move those into the source file, too, and I'm happy to do that.

**Comment by dschwen** [comment]
Are you talking about tutorial and examples under moose? I thought of them this time and they should be working fine. I do however have no access to any of the other apps. We might want to start thinking about getting me access. It definitely is handy to be able to do PRs for bison and the others I have access to.

File: framework/include/utils/InputParameters.h (hunk 1 above)
Comment (permcody) [inline framework/include/utils/InputParameters.h]: Oh you're right about the duplicate symbols, I forgot about that. Are these helpers called outside of the InputParameters class? If not and they can be moved to the .C file, let's do that instead.

**Comment by permcody** [comment]
Let's work with Tara and get all of the applications added to your security
plan. I'd rather just grant you full access to everything.

**Comment by amjokisaari** [comment]
@dschwen, this is nifty, what needs to be done on my end for hyrax?

**Comment by dschwen** [comment]
Just merge my PR once this moose PR is in

**Comment by amjokisaari** [comment]
great, thanks.

**Comment by dschwen** [comment]
Yak and Bison will be merged using the integration branch at [idaholab/moose/integration_5218](https://github.com/idaholab/moose/tree/integration_5218), so hopefully these get merged by tomorrow. @amjokisaari I just updated my hyrax branch with an integration branch submodule update, so it could be merged as well (MR opened).

**Comment by permcody** [comment]
Daniel, what in this PR causes an incompatible interface? I thought we'd be
able to make this change without the whole integration branch circus.

**Comment by dschwen** [comment]
It would have been possible to do this without an integration branch, but I
wanted to make use of the new way while we were making the changes. Yak for
example note pretty consistently uses MaterialPropertyName to declare input
parameters and used the new short form of get material property.

**Comment by dschwen** [comment]
Oh, and yak had its own typdef for MaterialPropertyName in a couple of
files. So rather than removing that we just went ahead to use the proper
type everywhere.

**Comment by friedmud** [comment]
So why is BISON failing to build?  This change shouldn't effect any of the current behavior of `getMaterialProperty()`.

**Comment by dschwen** [comment]
I had to one more getMaterialProperty const. The one that was overwritten
in Material.h

**Comment by friedmud** [comment]
Ah... makes sense.

**Comment by dschwen** [comment]
The MR to bison was merges 20mins ago.

File: framework/src/utils/InputParameters.C [hunk 3]
```diff
@@ -534,3 +541,119 @@ InputParameters::getDescription(const std::string & name)
...
+
+template<>
+void
+InputParameters::setParamHelper<MaterialPropertyName, int>(const std::string & /*name*/, MaterialPropertyName & l_value, const int &r_value)
+{
+  // Assign the default value so that it appears in the dump
+  std::ostringstream oss;
+  oss << r_value;
+  l_value = oss.str();
+}
```
Comment (permcody) [inline framework/src/utils/InputParameters.C]: This is a good change - thanks

**Comment by permcody** [comment]
@dschwen - This looks good, I'm going to merge this since it needs to go in next anyway. I'd appreciate if you write-up a blog post after we get everything integrated then we'll send the link out to the mailing list. This is a nice enhancement.

**Pull request merged**